src/merkle/simple_smt/mod.rs

```diff
@@ -0,0 +1,274 @@
+use super::{BTreeMap, Cow, Digest, Hasher, MerkleError, Vec};
+use core::{marker::PhantomData, mem::replace};
+
+mod path;
+pub use path::MerklePath;
+
+mod storage;
+pub use storage::{DepthDigest, MemoryTreeStorage, TreeStorage};
+
+#[cfg(test)]
+mod tests;
+
+// SPARSE MERKLE TREE
+// ================================================================================================
+
+/// A simple SMT backed by a hasher and tree storage.
+///
+/// This structure will not store any data, except the maximum length of the tree. All the data
+/// will be sent to its storage implementation.
+pub struct SparseMerkleTree<H, S>
+where
+    H: Hasher,
+    S: TreeStorage<Digest = H::Digest>,
+{
+    hasher: PhantomData<H>,
+    storage: S,
+    depth: u32,
+}
```

Comment:
I wonder if this should be moved to the storage too so the merkle tree is purely functional